core/src/hash/mod.rs

```diff
@@ -0,0 +1,444 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! High-level, built-in and efficient cryptographic hashing routines.
+
+mod hasher;
+
+use core::hash::Hash;
+
+use self::hasher::FinishInto;
+pub use self::hasher::{
+    Accumulator,
+    Blake2x128Hasher,
+    Blake2x256Hasher,
+    CryptoBuildHasher,
+    CryptoHasher,
+    Keccakx256Hasher,
+    Sha2x256Hasher,
+    TwoxHasher,
+    Wrap,
+};
+
+#[doc(inline)]
+pub use crate::env::hash::{
+    blake2_128 as blake2_128_raw_into,
+    blake2_256 as blake2_256_raw_into,
+    keccak_256 as keccak_256_raw_into,
+    sha2_256 as sha2_256_raw_into,
+    twox_128 as twox_128_raw_into,
+    twox_256 as twox_256_raw_into,
+    twox_64 as twox_64_raw_into,
+};
+
+macro_rules! impl_hash_fn_raw {
+    (
+        $(
+            $( #[$doc:meta] )*
+            fn $name:ident($output_len:literal);
+        )*
+    ) => {
+        $(
+            paste::item! {
+                $( #[$doc] )*
+                pub fn $name(input: &[u8]) -> [u8; $output_len] {
```

Comment:
IMO this is redundant. u8 slices implement `Hash` which means there is no reason to use this function the non-raw function. It would be nice to cut down the large number of functions.

Reply:
Yes I thought so as well but I was not sure exactly _HOW_ `u8` slices are hashes. E.g. are they hashed using their `len` + `hash(u8)` for each `u8` they contain just like a generic `[T]` slice or are `[u8]` actually specialized in NOT doing so. My fear is that there is no guarantee given by Rust how `[u8]` is really hashed and therefore we cannot get rid of the `_raw` APIs.

Reply:
Wouldn't you just use the `_raw_into` function it that rather low level use case? Does it really justify a new function? I am just genuinely asking. I feel like the number if overwhelming.